# 📊 Аналитический проект по исследованию игровой индустрии в начале  XXI века

## Цель проекта

проведение исследования игровой индустрии для дальнейшей оценки и формирования стратегии привлечения аудитории к игре «Секреты Темнолесья».


## Задачи проекта:
 - выгрузить данные и дата сета и провести предварительное исследование;
 - оценить качество предоставленных данных;
 - провести предобработку данных, в том числе: скорректировать тип данных, проанализировать  пропуски, обработать дубликаты и так далее;
 - провести фильтрацию и кластеризацию данных;
 - оценить топ-7 платформ по количеству игр, выпущенных за весь актуальный период;
 - сделать аналитические выводы.
  

### Описание данных

Данные `/datasets/new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

## 🗂️ Содержимое проекта

- Загрузка данных и знакомство с ними;
- Предварительная обработка данных игровой индустрии:
  - Проверка на наличие ошибок и подготовка данных;
  - Коррекция названий столбцов датафрейма;
  - Проверка типов данных;
  - Поиск и обработка пропусков;
  - Определение и удаление явных и неявных дубликатов;
- Фильтрация и очистка данных;
- Классификация данных;
- Подведение итогов и формулирование выводов;

# Загрузка данных и знакомство с ними

In [1]:
# импортируем необходимые библиотеки
import pandas as pd
import numpy as np

In [2]:
# Выгрузим данные 
try:
    games = pd.read_csv('C:/Users/Ирэн/Documents/Обучение/Data Analis_Яндекс_Практикум/Дата_сеты/datasets/new_games.csv', sep = ',')
except:
    games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv',sep = ',')

In [3]:
# Выгружаем общую инфорормацию
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
games.head(10)

Name Platform  Year of Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

   NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0     41.36    28.96     3.77         8.45          76.0          8      E  
1     29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2     15.68    12.76     3.79         3.29          82.0        8.3      E  
3     15.61    10.93     3.28         2.95          80.0          8      E  
4     11.27     8.89    10.22         1.00           NaN        NaN    NaN  
5     23.20     2.26     4.22         0.58           NaN        NaN    NaN  
6     11.28     9.14      6.5         2.88          89.0        8.5      E  
7     13.96     9.18     2.93         2.84          58.0        6.6      E  
8     14.44     6.94      4.7         2.24          87.0        8.4      E  
9     26.93     0.63     0.28         0.47           NaN        NaN    NaN

In [5]:
games.tail(10)

Name Platform  \
16946                                           15 Days       PC   
16947                     Men in Black II: Alien Escape       GC   
16948                                  Aiyoku no Eustia      PSV   
16949                Woody Woodpecker in Crazy Castle 5      GBA   
16950  SCORE International Baja 1000: The Official Game      PS2   
16951                     Samurai Warriors: Sanada Maru      PS3   
16952                                  LMA Manager 2007     X360   
16953                           Haitaka no Psychedelica      PSV   
16954                                  Spirits & Spells      GBA   
16955                               Winning Post 8 2016      PSV   

       Year of Release       Genre  NA sales EU sales JP sales  Other sales  \
16946           2009.0   Adventure      0.00     0.01      0.0          0.0   
16947           2003.0     Shooter      0.01      0.0      0.0          0.0   
16948           2014.0        Misc      0.00      0.0     0.01          0.0   
16949           2002.0    Platform      0.01      0.0      0.0          0.0   
16950           2008.0      Racing      0.00      0.0      0.0          0.0   
16951           2016.0      Action      0.00      0.0     0.01          0.0   
16952           2006.0      Sports      0.00     0.01      0.0          0.0   
16953           2016.0   Adventure      0.00      0.0     0.01          0.0   
16954           2003.0    Platform      0.01      0.0      0.0          0.0   
16955           2016.0  Simulation      0.00      0.0     0.01          0.0   

       Critic Score User Score Rating  
16946          63.0        5.8    NaN  
16947           NaN        tbd      T  
16948           NaN        NaN    NaN  
16949           NaN        NaN    NaN  
16950           NaN        NaN    NaN  
16951           NaN        NaN    NaN  
16952           NaN        NaN    NaN  
16953           NaN        NaN    NaN  
16954           NaN        NaN    NaN  
16955           NaN        NaN    NaN

### Предварительные выводы:

- исходный датафрейм состоит из `11 столбцов` и `16956 строк`, в которых отражена ключевая информация об исследуемых данных, при этом стоит рассмотреть переименования части столбцов для повышения их информативности и удобоваримости;
- данные представлены в 2х форматах, а именно: `object и  float64`,  при этом отметим, что в дальнейшем стоит преобразовать формат части столбцов;
- в столбцах `Critic Score`, `User Score`, `Rating` доля пропусков достаточно большая свыше 40% в каждом столбце. Стоит отметить, что пропуски в столбцах `Critic Score`, `User Score`, `Rating` коррелируют между собой, поэтому возможно природа возникновения у них одна. В частности если посмотреть на выгрузку, то в основном это данные по более старым играм, которые были выпущены в период когда рейтингов не было, а также оценки критиков и пользователей возможно и не получали. Поэтому, следует провести в дальнейшем более тщательный анализ данных пропусков для того чтобы выявить их закономерность и в итоге или удалить, пропуски, или обработать их заполнить, игнорировать и так далее.


---

# Предварительная обработка данных игровой индустрии

## Проверка на наличие ошибок и подготовка данных


In [6]:
print("Исходные названия столбцов:")
print(games.columns)

Исходные названия столбцов:
Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


Все названия столбцов представлены в одном формате, однако стиль не совсем соответствует принятому стилю написания, поэтому целесообразно привести все столбцы к формату shake_case.

In [7]:
games.columns = games.columns.str.lower().str.replace(' ', '_')

Теперь все столбцы приведены к единому регистру и целесообразно далее проводить обработку данных.

## Проверка типов данных

In [8]:
# Выгружаем общую инфорормацию
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [9]:
games['year_of_release'].unique()

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

In [10]:
games['eu_sales'].unique()

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

In [11]:
games['jp_sales'].unique()

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

In [12]:
games['user_score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

**Анализ проблемы типов данных:** 

Данные представлены в двух основных форматах: `Числовой (float64)` и `Строковый (object)`. Однако некоторые поля содержат несовместимые типы данных, что вызывает трудности при обработке:

- Поля eu_sales и jp_sales: присутствуют строковые значения "unknown" среди числовых данных.
- Поле user_score: имеет пропуски и смешанный формат (строковые и числовые значения).
- Поле year_of_release: представлен в формате float64, хотя ожидается тип данных datetime или int.

**Причины проблем:**
- Автоматическая конвертация: Автоматическое преобразование типов данных при загрузке данных могло привести к ошибкам.
- Присутствие пропусков: Пропуски вызывают несоответствие типов данных.
- Разнообразие исходных форматов: Данные могли изначально поступать в разных форматах, что привело к объединению разнородных данных.

In [13]:
# изменим тип данных с object на float или int
for column in ['eu_sales','jp_sales','user_score', 'year_of_release']:
     games[column] = pd.to_numeric(games[column], errors='coerce')

In [14]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


In [15]:
games.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
1      3.58      6.81         0.77           NaN         NaN    NaN  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
4      8.89     10.22         1.00           NaN         NaN    NaN

Скорректировали типы данных по столбцам, потом стоит  `year_of_release`  после обработки пропусков можно привести данные к `int` при необходимости.

## Поиск и обработка пропусков


Зафиксируем количество строк до обработки пропусков и дубликатов в `games`.

In [16]:
games_copy = games.copy()
start_games_count = len(games_copy)
print(f'Длина до обработки:{start_games_count}')

Длина до обработки:16956


Посчитаем  абсолютное и относительное количество пропусков в `games`.

In [17]:
sum_miss_games = games.isna().sum()
print ('Сумма пропущенных значение')
print( sum_miss_games)

Сумма пропущенных значение
name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64


In [18]:
temp = games.copy()
display(pd.DataFrame(round((temp.isna().mean()*100),2), columns=['NaNs, %'])
        .style.format( '{:.2f}').background_gradient('coolwarm'))

**Рекомендации по обработке пропусков:**

*Малая доля пропусков (<2%)*. Для столбцов `name`, `year_of_release`, `genre` рекомендуется удалить строки с пропусками, так как их количество незначительно. Это предотвратит влияние незначительных пропущенных данных на итоговый анализ.

*Средние показатели продаж `(eu_sales, jp_sales)`*.Эти столбцы имеют малое число пропусков `(~менее 2%)` рекомендуется использовать медиану для заполнения пропусков, поскольку, она устойчива к выбросам, лучше отражает центральную тенденцию распределения.

*Для других столбцов* с большим числом пропусков следует провести дополнительное исследование перед принятием решения о заполнении или удалении пропусков, так как игнорирование или удаление больших объемов данных может существенно повлиять на качество анализа.


Напишем функцию котора непосрдесвтенно помогает обнаруживать и  заполнять пропуски исходя из нескольких параметров.


In [19]:
##Напишем функцию которая непосредственно помогает обнаруживать и  заполнять пропуски исходя из нескольких параметров.
def fill_missing(row, col_name):
    if pd.isna(row[col_name]):
        group= (games['platform'] == row['platform']) & (games['year_of_release'] == row['year_of_release'])
        mean_val = games[group][col_name].median()
        return mean_val
    else:
        return row[col_name]

# Применение функции для заполнения пропусков
for col in ['eu_sales', 'jp_sales']:
    games[col] = games.apply(lambda row: fill_missing(row, col), axis=1)

In [20]:
games.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
1      3.58      6.81         0.77           NaN         NaN    NaN  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
4      8.89     10.22         1.00           NaN         NaN    NaN

In [21]:
games.describe()

year_of_release      na_sales      eu_sales      jp_sales  \
count     16681.000000  16956.000000  16956.000000  16956.000000   
mean       2006.485522      0.262023      0.144078      0.077152   
std           5.873102      0.808654      0.500221      0.307114   
min        1980.000000      0.000000      0.000000      0.000000   
25%        2003.000000      0.000000      0.000000      0.000000   
50%        2007.000000      0.080000      0.020000      0.000000   
75%        2010.000000      0.240000      0.110000      0.040000   
max        2016.000000     41.360000     28.960000     10.220000   

        other_sales  critic_score   user_score  
count  16956.000000   8242.000000  7688.000000  
mean       0.047087     68.926717     7.122607  
std        0.185577     13.944565     1.499804  
min        0.000000     13.000000     0.000000  
25%        0.000000     60.000000     6.400000  
50%        0.010000     71.000000     7.500000  
75%        0.030000     79.000000     8.200000  
max       10.570000     98.000000     9.700000

In [22]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  float64
 6   jp_sales         16956 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


In [23]:
games = games.dropna(subset=['name','year_of_release','genre']).reset_index(drop=True)

In [24]:
print ('Доля пропущенных значение')
print(100*games.isna().mean())

Доля пропущенных значение
name                0.000000
platform            0.000000
year_of_release     0.000000
genre               0.000000
na_sales            0.000000
eu_sales            0.000000
jp_sales            0.000000
other_sales         0.000000
critic_score       51.525871
user_score         54.685533
rating             40.637928
dtype: float64


In [25]:
##приведем 'year_of_release' к int для удобства анализа
games['year_of_release'] = games['year_of_release'].astype('int64')

В обновленных данных пропуски отсутствуют по обработанным столбцам, поэтому, далее целесообразно провести исследование других столбцов с пропусками более подробно.

In [26]:
games.describe()

year_of_release      na_sales      eu_sales      jp_sales  \
count     16679.000000  16679.000000  16679.000000  16679.000000   
mean       2006.487140      0.262685      0.144951      0.078023   
std           5.871597      0.813443      0.503607      0.309367   
min        1980.000000      0.000000      0.000000      0.000000   
25%        2003.000000      0.000000      0.000000      0.000000   
50%        2007.000000      0.080000      0.020000      0.000000   
75%        2010.000000      0.240000      0.110000      0.040000   
max        2016.000000     41.360000     28.960000     10.220000   

        other_sales  critic_score   user_score  
count  16679.000000   8085.000000  7558.000000  
mean       0.047342     68.958689     7.124438  
std        0.186854     13.923424     1.498803  
min        0.000000     13.000000     0.000000  
25%        0.000000     60.000000     6.400000  
50%        0.010000     71.000000     7.500000  
75%        0.030000     79.000000     8.200000  
max       10.570000     98.000000     9.700000

**Аномалии в данных:** Количественные показатели содержат аномальные отклонения, вероятно обусловленные временным диапазоном данных (начиная с 1980-х гг.) и стремительным развитием игровой индустрии с 2000-х годов благодаря развитию интернета и технологий.

**Доля пропусков:** В столбце `rating` доля пропусков достигает `40,64%`, что значительно влияет на полноту данных. Удаление такого большого количества пропусков приведет к потере значимой части данных и ухудшению результатов анализа. Поэтому предпочтительно сохранить данные и применить методы восстановления пропусков.

**Корреляция пропусков:** Пропуски в `rating` коррелируют с пропусками в `critic_score` и `user_score`, что предполагает единую природу возникновения пропусков. Следовательно, подходы к обработке пропусков в этих трех столбцах могут быть аналогичными.

Исследуем  какие рейтинги представлены в данных `rating`

In [27]:
print ('Доля пропущенных значение rating')
print(100*games['rating'].isna().mean())

Доля пропущенных значение rating
40.63792793332934


In [28]:
games['rating'].value_counts()

E       3968
T       2948
M       1558
E10+    1414
EC         8
K-A        3
AO         1
RP         1
Name: rating, dtype: int64

Больше всего заполненных данных присутствует в рейтингах `E`,`T`,`M`,`E10+`, это игры по сути для аудитории без ограничения, так как они более популярны и доступны как детям, так и взрослым, а следовательно, их покупают чаще всего.
Отметим, что у нас есть рейтинг `K-A`, который с 1998 года был переименован в `E`. Поэтому, для оптимизации анализа может переименовать данные в строках

In [29]:
games['rating'] = games['rating'].replace('K-A','E')
games['rating'].value_counts()

E       3971
T       2948
M       1558
E10+    1414
EC         8
AO         1
RP         1
Name: rating, dtype: int64

In [30]:
games['year_of_release'].value_counts()

2009    1450
2008    1445
2010    1279
2007    1218
2011    1149
2006    1020
2005     950
2002     839
2003     789
2004     771
2012     670
2015     612
2014     584
2013     552
2016     507
2001     491
1998     384
2000     357
1999     341
1997     293
1996     267
1995     220
1994     121
1993      60
1981      46
1992      43
1991      42
1982      37
1986      22
1983      18
1987      17
1989      17
1990      16
1988      15
1985      14
1984      14
1980       9
Name: year_of_release, dtype: int64

Игры, выпущенные до `1994 года`, часто не имели официальных рейтингов, что объясняет наличие пропусков в данных. Эти пропуски являются нормальным явлением и не требуют исправления.

Поскольку пропуски распространены повсеместно (во всех периодах и жанрах), простое заполнение средних значениями невозможно. Вместо этого предлагается использовать специальное обозначение `(«заглушку») — unknown`, что позволит сохранить структуру данных и продолжить дальнейший анализ.

In [31]:
games['rating'] = games['rating'].fillna('unknown')
games['rating'].value_counts()

unknown    6778
E          3971
T          2948
M          1558
E10+       1414
EC            8
AO            1
RP            1
Name: rating, dtype: int64

Исследуем  `critic_score` на наличие пропусков и проведем их обработку.

In [32]:
print ('Доля пропущенных значение critic_score')
print(100*games['critic_score'].isna().mean())

Доля пропущенных значение critic_score
51.52587085556688


Пропущенных значений достаточно много, соответственно удалять их не следует так как это может негативно сказаться на результатах исследования так как выборка сократиться более чем на `50%`. Поэтому стоит проводить замену пропусков, однако взять среднее значение или медианное мы не можем, так как все- таки `critic_score` - это скорее качественный показатель, нежели количественный.
Поэтому, будем использовать `заглушки`, только уже числовые и возьмем за основу `-10`, так как все что более или равно `0` может быть воспринято как реальные данные.


In [33]:
games['critic_score'] = games['critic_score'].fillna('-10')

## приведем еще и к необходимому типу данных int64
games['critic_score'] = games['critic_score'].astype('int64')

In [34]:
print(100*games['critic_score'].isna().mean())

0.0


Пропусков в `critic_score` нет, поэтому можем далее обрабатывать данные. 

Исследуем `user_score` на наличие пропусков.

In [35]:
print ('Доля пропущенных значение user_score')
print(100*games['user_score'].isna().mean())

Доля пропущенных значение user_score
54.68553270579771


Причина пропуска оценок критиками `(critic_score)` и пользователями `(user_score)`, составляющих более 50%, может заключаться в следующем:

- Оценки могли не проводиться ввиду низкой популярности игры.
- Технические проблемы или сбой системы оценки.
- Игры выходили одновременно с крупными релизами на популярных платформах.
- Из-за высокого процента пропусков их удаление недопустимо, иначе данные будут сильно искажены. Оба столбца тесно связаны друг с другом, поэтому обработку пропусков лучше проводить одинаково.


In [36]:
games['user_score'] = games['user_score'].fillna('-1')
## приведем еще и к необходимому типу данных
games['user_score'] = games['user_score'].astype('float64')

In [37]:
print(100*games['user_score'].isna().mean())

0.0


Выгрузим сводную инфорамци после обработк пропусков.

In [38]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16679 entries, 0 to 16678
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16679 non-null  float64
 6   jp_sales         16679 non-null  float64
 7   other_sales      16679 non-null  float64
 8   critic_score     16679 non-null  int64  
 9   user_score       16679 non-null  float64
 10  rating           16679 non-null  object 
dtypes: float64(5), int64(2), object(4)
memory usage: 1.4+ MB


**Объем пропусков варьируется от 0,1% до 50%, при этом среди основных причин можно выделить:**

- Ошибки при выгрузке данных.
- Отсутствие данных по объективным причинам.

В категоричных столбцах `(name, year_of_release, genre)` доля пропусков: не более 2%. Данные пропуски были  удаленны, так как нет оптимального подхода к их заполнению. а удаление их приведет к незначительной утрате.

В количественных столбцах `(eu_sales, jp_sales)` с долей пропусков до 2% было принято решение о заполнить  пропуски медианой  для минимизации влияния выбросов.

В столбцах с критериями качества `(critic_score, user_score, rating)` - доля пропусков более 40%. Поэтому, пропуски были промаркированы специальной метокой-заглушкой `(unknown)`, так как полное восстановление данных трудоемко и неэффективно.
Использование заглушек позволяет сохранить данные для дальнейшего анализа, несмотря на значительный процент пропусков.


## Определение и удаление явных и неявных дубликатов

Изучим категорийные данные и начнем с `name`.

In [39]:
print('Количество уникальных значений наименований игр')
games['name'].nunique()

Количество уникальных значений наименований игр


11426

Приведем наименования к единому регистру и удалим возможные пропуски.

In [40]:
games['name'] = games['name'].str.lower().str.strip()
print('Количество уникальных значений наименований игр после обработки')
print(games['name'].nunique())

Количество уникальных значений наименований игр после обработки
11426


Изучим  данные в столбце `genre`.

In [41]:
print('Наименование жанров')
games['genre'].unique()

Наименование жанров


array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [42]:
print('Количевство жанров')
print(games['genre'].nunique())

Количевство жанров
24


In [43]:
games['genre'] = games['genre'].str.lower()
games['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

In [44]:
print('Количевство жанров после обработки:')
print(games['genre'].nunique())

Количевство жанров после обработки:
12


Исследуем игровые платформы.

In [45]:
print('Уникальные наименования платформ:')
games['platform'].sort_values().unique()

Уникальные наименования платформ:


array(['2600', '3DO', '3DS', 'DC', 'DS', 'GB', 'GBA', 'GC', 'GEN', 'GG',
       'N64', 'NES', 'NG', 'PC', 'PCFX', 'PS', 'PS2', 'PS3', 'PS4', 'PSP',
       'PSV', 'SAT', 'SCD', 'SNES', 'TG16', 'WS', 'Wii', 'WiiU', 'X360',
       'XB', 'XOne'], dtype=object)

Как видно из данных здесь дублей нет, все платформы являются уникальными, поэтому обработки не требуется.

In [46]:
print('Уникальные наименования рейтингов:')
games['rating'].sort_values().unique()

Уникальные наименования рейтингов:


array(['AO', 'E', 'E10+', 'EC', 'M', 'RP', 'T', 'unknown'], dtype=object)

In [47]:
print('Данные по годам реализов:')
games['year_of_release'].sort_values().unique()

Данные по годам реализов:


array([1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990,
       1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001,
       2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012,
       2013, 2014, 2015, 2016], dtype=int64)

Мы провели обработку датафрейма от неявных дубликатов, что позволит нам упростить  работу в части обработки уже явных дубликатов

In [48]:
print('Количество явных дубликатов в датасете')
games_duplicated=games.duplicated().sum()

Количество явных дубликатов в датасете


In [49]:
games = games.drop_duplicates(subset=None, keep='first', inplace=False) 

In [50]:
games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16444 entries, 0 to 16678
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  int64  
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float64
 6   jp_sales         16444 non-null  float64
 7   other_sales      16444 non-null  float64
 8   critic_score     16444 non-null  int64  
 9   user_score       16444 non-null  float64
 10  rating           16444 non-null  object 
dtypes: float64(5), int64(2), object(4)
memory usage: 1.5+ MB


Обнаружено дублирование данных в столбце` genre`, вызванное различиями в регистре данных. Для устранения этой проблемы применено приведение всех записей к одному регистру, что гарантирует устранение потенциальных дубликатов.

Проверка других качественных столбцов показала отсутствие дублирующих значений.

Используя метод `drop_duplicates()`, устранено `235 явных дубликатов`. Итоговая таблица включает `11 столбцов` и `16444 строки`.

После обработки данных стоит посмотреть какая доля данных была утеряна в итоге удаления дубликатов, пропусков и так далее.

In [51]:
end_games_count = len(games)
print(f'Длина датафрейма после обработки: {end_games_count}')

Длина датафрейма после обработки: 16444


In [52]:
# Проверим сколько удалено строк датасета

print(" Было строк в исходном датасете", start_games_count,
      '\n', "Осталось строк в датасете после обработки", end_games_count,
      '\n', "Удалено строк в датасете после обработки", start_games_count-end_games_count,
      '\n', "Процент потерь", round((start_games_count-end_games_count)/end_games_count*100, 2))

 Было строк в исходном датасете 16956 
 Осталось строк в датасете после обработки 16444 
 Удалено строк в датасете после обработки 512 
 Процент потерь 3.11


## Предварительные выводы по разделу:

На основании проведенных расчетов можно сделать предварительный вывод о том, что:
-  в процессе обработки данных все наименования столбцов были приведены к единому формату для оптимизации процесса анализа и понимания информации;
- данные в `games` были приведены к корректному типу данных, чтобы было в дальнейший корректный анализ и исследования;
- в процессе обработки были выявлены пропуски в стобцах `name`,`year_of_release`,`genre`,`critic_score`,`user_score`, `rating`,`eu_sales`, `jp_sales`;
- пропуски в `name`,`year_of_release`,`genre` были удалены, так как их доля не превышает `2%`, поэтому, их удаление значительно не повлияет на данные;
- пропуски в `eu_sales`, `jp_sales` были заменены на медианные значения исходя их года реализации и платформы, это было необходимо чтобы данные были более реалистичными, однако здесь данные также можно было удалить, так как доля в общем объеме данных составляла менее 2%;
- пропуски в столбцах `critic_score`,`user_score`, `rating` были заменены так как доля пропусков составляла более `40%` в каждом столбце,  что делает их удаление не целесообразным, поэтому пропуски были заменены на `заглушки`;
- были удалены явные и неявные дубликаты по средствам выявления задвоенных данных, некорректно внесенных показателей и так далее;  
- до обработки данных `games` состоял из `16956 строк` и `11 столбцов`, при этом было удаленно `512 строк`,  то есть `3%` от всего  `games`  , то есть  основная часть данных сохранилась, следовательно, это позволит нам сделать более тщательный и валидный анализ.


---

#  Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберем данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [53]:
df_actual = games.copy()

In [54]:
df_actual.head()

name platform  year_of_release         genre  na_sales  \
0                wii sports      Wii             2006        sports     41.36   
1         super mario bros.      NES             1985      platform     29.08   
2            mario kart wii      Wii             2008        racing     15.68   
3         wii sports resort      Wii             2009        sports     15.61   
4  pokemon red/pokemon blue       GB             1996  role-playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score   rating  
0     28.96      3.77         8.45            76         8.0        E  
1      3.58      6.81         0.77           -10        -1.0  unknown  
2     12.76      3.79         3.29            82         8.3        E  
3     10.93      3.28         2.95            80         8.0        E  
4      8.89     10.22         1.00           -10        -1.0  unknown

In [55]:
games_actual = df_actual.loc[(df_actual['year_of_release'] >= 2000) & (df_actual['year_of_release'] <= 2013)].copy()
games_actual.reset_index(drop=True, inplace=True)

In [56]:
games_actual.head()

name platform  year_of_release     genre  na_sales  \
0             wii sports      Wii             2006    sports     41.36   
1         mario kart wii      Wii             2008    racing     15.68   
2      wii sports resort      Wii             2009    sports     15.61   
3  new super mario bros.       DS             2006  platform     11.28   
4               wii play      Wii             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45            76         8.0      E  
1     12.76      3.79         3.29            82         8.3      E  
2     10.93      3.28         2.95            80         8.0      E  
3      9.14      6.50         2.88            89         8.5      E  
4      9.18      2.93         2.84            58         6.6      E

Проверим корректно ли выгрузились данные и для этого найдем минимальное и максимальое значение `year_of_release`.

In [57]:
## Проверим корректно ли выгрузились данные и для этого найдем минимальное и максимальое значение `year_of_release`.
print('Максимальное значение в столбце')
games_actual['year_of_release'].max()

print('Минимальное значение в столбце')
games_actual['year_of_release'].min()

Максимальное значение в столбце
Минимальное значение в столбце


2000

---

# Категоризация данных
    
Проведите категоризацию данных: разделим все игры по оценкам пользователей и выделите такие категории: 
- `высокая оценка` (от 8 до 10 включительно);
- `средняя оценка` (от 3 до 8, не включая правую границу интервала);
- `низкая оценка` (от 0 до 3, не включая правую границу интервала).

Проверим уникальные значение в столбце `user_score`.

In [58]:
#Проверим уникальные значение в столбце `user_score`.
games_actual['user_score'].sort_values().unique()

array([-1. ,  0. ,  0.2,  0.3,  0.5,  0.6,  0.7,  0.9,  1. ,  1.1,  1.2,
        1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,  2.3,
        2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,  3.4,
        3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,  4.5,
        4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,  5.6,
        5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,  6.7,
        6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,  7.8,
        7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,  8.9,
        9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7])

Мы видим, что оценки пользователей находятся в диапазоне `от -1 до 9.7`, где необходимо помнить, что `-1` - это `наша заглушка`, то есть  игры без оценок, поэтому целесообразно их указать как отдельный тип `Без оценки`, чтобы данные не влияли на общую картину.

In [59]:
bins = [float('-inf'), 0.0, 3.0, 8.0, 11]
labels = ['Без оценки', 'Низкая оценка', 'Средняя оценка', 'Высокая оценка']

# Категоризация оценок
games_actual['user_score_categories'] = pd.cut(games_actual['user_score'], bins=bins, labels=labels, right=False)


In [60]:
# Проверяем границы категорий по заданию брифа и созданные в проекте
tmp0, u_s, c_s = games_actual.copy(), 'user_score', 'critic_score'
cat_u, cat_c = 'user_score_categories', 'critic_score_categories'
print("Оценка 8 должна относиться категории 'высокая оценка'", "здесь это",
      tmp0[tmp0[u_s] == 8][cat_u].unique())
print("Оценка 3 должна относиться категории 'средняя оценка'", "здесь это",
      tmp0[tmp0[u_s] == 3][cat_u].unique())
print("Оценка 0 должна относиться категории 'низкая оценка'", "здесь это",
      tmp0[tmp0[u_s] == 0][cat_u].unique())

Оценка 8 должна относиться категории 'высокая оценка' здесь это ['Высокая оценка']
Categories (4, object): ['Без оценки' < 'Низкая оценка' < 'Средняя оценка' < 'Высокая оценка']
Оценка 3 должна относиться категории 'средняя оценка' здесь это ['Средняя оценка']
Categories (4, object): ['Без оценки' < 'Низкая оценка' < 'Средняя оценка' < 'Высокая оценка']
Оценка 0 должна относиться категории 'низкая оценка' здесь это ['Низкая оценка']
Categories (4, object): ['Без оценки' < 'Низкая оценка' < 'Средняя оценка' < 'Высокая оценка']


In [61]:
games_actual.head()

name platform  year_of_release     genre  na_sales  \
0             wii sports      Wii             2006    sports     41.36   
1         mario kart wii      Wii             2008    racing     15.68   
2      wii sports resort      Wii             2009    sports     15.61   
3  new super mario bros.       DS             2006  platform     11.28   
4               wii play      Wii             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0     28.96      3.77         8.45            76         8.0      E   
1     12.76      3.79         3.29            82         8.3      E   
2     10.93      3.28         2.95            80         8.0      E   
3      9.14      6.50         2.88            89         8.5      E   
4      9.18      2.93         2.84            58         6.6      E   

  user_score_categories  
0        Высокая оценка  
1        Высокая оценка  
2        Высокая оценка  
3        Высокая оценка  
4        Средняя оценка

Проверим уникальные значение в столбце `critic_score` для определения того корректно ли поставлены категории распределения.

In [62]:
games_actual['critic_score'].sort_values().unique()

array([-10,  13,  17,  19,  20,  21,  23,  24,  25,  26,  27,  28,  29,
        30,  31,  32,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,
        43,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,
        56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,
        69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,
        82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,
        95,  96,  97,  98], dtype=int64)

Разделим все игры по оценкам критиков и выделите такие категории: 
- высокая оценка (от 80 до 100 включительно);
- средняя оценка (от 30 до 80, не включая правую границу интервала);
- низкая оценка (от 0 до 30, не включая правую границу интервала).

In [63]:
bins = [-10,0,30, 80, 101]
labels = ['Без оценки','Низкая оценка', 'Средняя оценка', 'Высокая оценка']

# Применяем метод cut с параметром right=False
games_actual['critic_score_categories'] = pd.cut(games_actual['critic_score'], bins = bins, labels = labels,right=False)

In [64]:
# Проверяем границы категорий по заданию брифа и созданные в проекте
tmp0, u_s, c_s = games_actual.copy(), 'user_score', 'critic_score'
cat_u, cat_c = 'user_score_categories', 'critic_score_categories'
print("Оценка 80 должна относиться категории 'высокая оценка'", "здесь это",
      tmp0[tmp0[c_s] == 80][cat_c].unique())
print("Оценка 30 должна относиться категории 'средняя оценка'", "здесь это",
      tmp0[tmp0[c_s] == 30][cat_c].unique())
print("Оценка 0 должна относиться категории 'низкая оценка'", "здесь это",
      tmp0[tmp0[c_s] == 0][cat_c].unique())

Оценка 80 должна относиться категории 'высокая оценка' здесь это ['Высокая оценка']
Categories (4, object): ['Без оценки' < 'Низкая оценка' < 'Средняя оценка' < 'Высокая оценка']
Оценка 30 должна относиться категории 'средняя оценка' здесь это ['Средняя оценка']
Categories (4, object): ['Без оценки' < 'Низкая оценка' < 'Средняя оценка' < 'Высокая оценка']
Оценка 0 должна относиться категории 'низкая оценка' здесь это [], Categories (4, object): ['Без оценки' < 'Низкая оценка' < 'Средняя оценка' < 'Высокая оценка']


Проверяем как отобразились данные в датафрейсе.

In [65]:
games_actual.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12781 entries, 0 to 12780
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   name                     12781 non-null  object  
 1   platform                 12781 non-null  object  
 2   year_of_release          12781 non-null  int64   
 3   genre                    12781 non-null  object  
 4   na_sales                 12781 non-null  float64 
 5   eu_sales                 12781 non-null  float64 
 6   jp_sales                 12781 non-null  float64 
 7   other_sales              12781 non-null  float64 
 8   critic_score             12781 non-null  int64   
 9   user_score               12781 non-null  float64 
 10  rating                   12781 non-null  object  
 11  user_score_categories    12781 non-null  category
 12  critic_score_categories  12781 non-null  category
dtypes: category(2), float64(5), int64(2), object(4)
memory usage:

In [66]:
games_actual.head(20)


name platform  year_of_release  \
0                                     wii sports      Wii             2006   
1                                 mario kart wii      Wii             2008   
2                              wii sports resort      Wii             2009   
3                          new super mario bros.       DS             2006   
4                                       wii play      Wii             2006   
5                      new super mario bros. wii      Wii             2009   
6                                     nintendogs       DS             2005   
7                                  mario kart ds       DS             2005   
8                                        wii fit      Wii             2007   
9                             kinect adventures!     X360             2010   
10                                  wii fit plus      Wii             2009   
11                            grand theft auto v      PS3             2013   
12                 grand theft auto: san andreas      PS2             2004   
13  brain age: train your brain in minutes a day       DS             2005   
14                 pokemon diamond/pokemon pearl       DS             2006   
15                            grand theft auto v     X360             2013   
16                   grand theft auto: vice city      PS2             2002   
17                 pokemon ruby/pokemon sapphire      GBA             2002   
18   brain age 2: more training in minutes a day       DS             2005   
19                   pokemon black/pokemon white       DS             2010   

           genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
0         sports     41.36     28.96      3.77         8.45            76   
1         racing     15.68     12.76      3.79         3.29            82   
2         sports     15.61     10.93      3.28         2.95            80   
3       platform     11.28      9.14      6.50         2.88            89   
4           misc     13.96      9.18      2.93         2.84            58   
5       platform     14.44      6.94      4.70         2.24            87   
6     simulation      9.05     10.95      1.93         2.74           -10   
7         racing      9.71      7.47      4.13         1.90            91   
8         sports      8.92      8.03      3.60         2.15            80   
9           misc     15.00      4.89      0.24         1.69            61   
10        sports      9.01      8.49      2.53         1.77            80   
11        action      7.02      9.09      0.98         3.96            97   
12        action      9.43      0.40      0.41        10.57            95   
13          misc      4.74      9.20      4.16         2.04            77   
14  role-playing      6.38      4.46      6.04         1.36           -10   
15        action      9.66      5.14      0.06         1.41            97   
16        action      8.41      5.49      0.47         1.78            95   
17  role-playing      6.06      3.90      5.38         0.50           -10   
18        puzzle      3.43      5.35      5.32         1.18            77   
19  role-playing      5.51      3.17      5.65         0.80           -10   

    user_score   rating user_score_categories critic_score_categories  
0          8.0        E        Высокая оценка          Средняя оценка  
1          8.3        E        Высокая оценка          Высокая оценка  
2          8.0        E        Высокая оценка          Высокая оценка  
3          8.5        E        Высокая оценка          Высокая оценка  
4          6.6        E        Средняя оценка          Средняя оценка  
5          8.4        E        Высокая оценка          Высокая оценка  
6         -1.0  unknown            Без оценки              Без оценки  
7          8.6        E        Высокая оценка          Высокая оценка  
8          7.7        E        Средняя оценка          Высокая оценка  
9          6.3        E        Средняя оценка          Средняя оцен

Все данные заполненны,  пропуски отсутвуют и данные выгружены в корректном типе.

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [67]:
games_actual.groupby(['user_score_categories'])['name'].count()


user_score_categories
Без оценки        6298
Низкая оценка      116
Средняя оценка    4081
Высокая оценка    2286
Name: name, dtype: int64

In [68]:
grouped_games = games_actual.groupby(['critic_score_categories'])['name'].count()
grouped_games

critic_score_categories
Без оценки        5612
Низкая оценка       55
Средняя оценка    5422
Высокая оценка    1692
Name: name, dtype: int64

Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [69]:
top_7_games = games_actual.groupby('platform')['name'].agg('count').sort_values(ascending=False).head(7)
print('Топ - 7 платформ по количеству игр\n')
print(top_7_games)

Топ - 7 платформ по количеству игр

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: name, dtype: int64


Таким образом, по результату исследования мы видим, что наиболее популярными в период с 2000 по 2013 год были игры для платформ `PS2,DS,Wii, PSP, X360,PS3,GBA`, отметим также что в основном это игры  как для полноразмерных игровых консолей, так и для портативных. Однако добавим дополнительно информацию и по расчету сумм продаж в зависимости от региона, при этом за точку сортировки пример `na_sales`.

In [70]:
top_games_sum = games_actual.groupby('platform').agg({'na_sales':'sum','eu_sales' : 'sum','jp_sales':'sum','other_sales':'sum'})
sorted_top_games = top_games_sum.sort_values(by = 'na_sales',ascending=False)
print('Топ - 7 платформ по стоимости продаж игр\n')
print(sorted_top_games.head(7))

Топ - 7 платформ по стоимости продаж игр

          na_sales  eu_sales  jp_sales  other_sales
platform                                           
PS2         572.92    331.83    137.54       190.47
X360        567.44    252.20     12.20        80.53
Wii         484.79    255.73     68.28        77.31
DS          380.31    187.29    175.00        59.04
PS3         368.08    301.67     68.19       125.22
GBA         184.12     74.12     46.56         7.61
XB          182.06     59.65      1.38         8.48


Отметим, что по количеству, что по сумам продаж `PS2` во всем мире, кроме японского региона. Там все - таки более популярны игр для консолей `Nintendo`,  и этому есть объективное обоснование, так как Япония все - таки родина для платформ связанных с играми `Nintendo`.

---

# Подведение итогов и формулирование выводов

### Основные этапы проведенной работы.

Работа посвящена исследованию игровой индустрии на основании данных за период с `1980 по 2016 год`, охватывая ключевые изменения в развитии отрасли. Активный рост индустрии пришелся на начало 2000-х годов, связанный с ростом доступности компьютеров, игровых приставок и распространением интернета.


**1. Преобразование структуры данных:**
- Переименованы столбцы в единый стиль snake_case;
- Изменены типы данных ряда столбцов на float, включая `eu_sales`, `jp_sales`, `user_score`, `year_of_release`, что обеспечило возможность дальнейших расчетов.

**2. Очистка данных:**
- Удалены строки с небольшими долями пропусков (столбцы `name`, `year_of_release`, `genre`).
- Пропуски в столбцах `eu_sales`, `jp_sales` заменены медианными значениями.
- Строки с пропусками в критически важных полях `(critic_score, user_score, rating)` отмечены специальными значениями ("заглушками").

**3. Удаление дубликатов:**
- Устранены `235 явных дубликатов` и скрытые повторения путем нормализации формата данных.

**4. Подготовка набора данных:**
- Проведен срез данных по годам `(2000—2013)`, соответствующий активному росту игровой индустрии.
- Созданы новые категориальные переменные на основе среднего балла критики и пользователей.
- Выполнена агрегация данных по платформам, выявив топ-7 платформ с максимальным количеством игр и продаж.



Основная часть анализа сосредоточена на периоде `2000—2013 гг`. Именно этот промежуток представляет наиболее полный и информативный массив данных.

Оценки критиков и пользователей играют важную роль в определении успеха игры. Однако значительная часть игр не имеет отзывов либо страдает от технических ошибок в оценке. Таким образом, около половины игр не имеют четких оценок ни от пользователей, ни от критиков, что ограничивает возможности анализа.

Распределение оценок выглядит следующим образом:

- Средняя оценка: 4081 игра.
- Высокая оценка (более 8): 2286 игр.
- Низкая оценка: 116 игр.

Большинство игр получило средние оценки (от 30 до 80 баллов), высокие оценки присуждены лишь немногим. Из-за частых пропусков обе группы оценок сильно коррелируют друг с другом.

Самые популярные игровые платформы периода исследований включают:

- PlayStation 2 (PS2),
- DS,
- Wii,
- Xbox 360 (X360),
- PlayStation 3 (PS3),
- Game Boy Advance (GBA).

Популярность `PS 2` объясняется её новаторством и массовой популярностью среди геймеров. Платформы `Sony` и `Microsoft` доминировали на рынках Северной Америки и Европы, в то время как продукция Nintendo оставалась лидером в Японии.

Успех игры определяется различными факторами, такими как жанр, оценки пользователей и особенности регионов реализации.